sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc

```diff
@@ -114,17 +114,35 @@ supports.
  feature-test macro always has this value.
 |===
 
+=== Kernel Property Trait
+
+Some kernel properties carry compile-time information to affect how the compiler
+handles the corresponding kernel object. However, some of the kernel enqueue
+functions take a `sycl::kernel` object and as such it may not be possible to
+infer which kernel function it refers to at compile-time. To help identify the
+properties that have compile-time effects on the kernel, the
+`has_compile_time_kernel_effect` trait.
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+template<typename> struct has_compile_time_kernel_effect : std::false_type {};
+
+template<typename PropertyT>
+static constexpr bool has_compile_time_kernel_effect_v =
+    has_compile_time_kernel_effect<PropertyT>::value;
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
```

Comment:
I would rather not expose this as a new API.  I've been thinking instead that we should extend the meaning of `is_property_key_of` to allow arbitrary classification of properties.  For example, we could define two types like this:

```
struct kernel_properties;
struct kernel_launch_properties;
```

We would then document each kernel property indicating which classification it has: `kernel_property` vs. `kernel_launch_property`.  We could then define constraints for APIs that take a property list using something like `is_property_key_of<Prop, kernel_launch_properties>`.  Application code could use this trait for its own constraints.

There is still some name bikeshedding to do.  For example, are these classification types directly in `sycl::ext::oneapi::experimental` or are they in a sub-namespace?  I think there should also be some sort of trait to tell if all properties in a list have a particular classification (e.g. `is_property_list_key_of<PropertyListT, kernel_launch_properties>`.  This name could also use some bikeshedding.

I think there are a couple options for this PR.  You could push the ideas I outline above forward into something more concrete, and then change this PR to use those new APIs.

Or, you could simply remove `has_compile_time_kernel_effect` from this PR and and use less precise language about the constraints on the `single_task`. etc. functions below.  For example, the constraint can just say that the properties must be launch properties without defining a specific trait.  This would allow us to make the language more precise later once we refine the property classification traits.  In this case, the implementation could still define traits in the `detail` namespace, but they would be undocumented.

Reply:
I went with the latter to buy us time so we can settle on a good solution here.